### DISCUSSION TITLE
Performance considerations vs drehmflight

### DISCUSSION THREAD

**Discussion by jonbjornn** [qqqlab/madflight#25] [Q&A]
*Started this originally as an issues but then noticed there were active discussions!*

Love the fact that this supports a wider range of hardware, but wondered what compromises had to be made to support such drastically different mcu clock speeds?

What do you give up in comparison to drehmflight in order to gain the broader hw compatibility?

**Comment by qqqlab** [comment]
Actually not many compromises where made, I think the performance between drehmflight and madflight will be similar. The different hardware platforms in madflight are implemented with #defines, and don't affect runtime.

The drehmfight target teensy 4.0 runs at 600MHz, whereas most madflight targets run at less than half that clock speed.

Drehmflight uses software oneshot125 to drive the ESCs, which introduces a wait of up to 125us in each 500us loop, i.e. up to 25% of computing power is wasted. madflight uses hardware pwm to drive the ESCs.

Neither drehmflight nor madflight use DMA, which would probably be the first thing to implement if somebody is really concerned about performance.

**Comment by Blimpyway** [comment]
Hi, a few observations on RP2350/RP2040 here, 
- There is a huge potential to overclock, e.g. [in this thread](https://forums.raspberrypi.com/viewtopic.php?t=375087) they compare rp2350 clocked at 500-600Mhz out-matching a  STM32H743/753  at 500Mhz
- The PIO can offload the uC cores from time sensitive I/O, e.g. I could not find libs for multishot125 but[ there are some for DSHOT](https://github.com/josephduchesne/pico-dshot-bidir) .

**Reply by qqqlab** [reply]
Even without overclocking the 150MHz RP2350 is a powerhouse, and is together with the ESP32-S3 my favorite target. Raspberry Pi apparently recently updated the certified clock speed of RP2040 from 133 to 200 MHz, maybe the same happens to RP2350.

DSHOT would be a nice addition, although not high on my personal [wish list](https://github.com/qqqlab/madflight/discussions/42). But feel free to submit a PR for it and I'll merge it.
